### PR TITLE
Initial commit for web UI/API

### PR DESCRIPTION
**Pull request by tadhg-ohiggins**
Web UI first pull
Django REST framework magic
Sadly no tests yet

### CODE REVIEW

**Comment by tadhg-ohiggins** [comment]
This isn't an urgent merge—a PR is the easiest way to share what I've been working on.

File: regparser/web/jobs/models.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from django.db import models
+
+
+class ParsingJob(models.Model):
+
```
Comment (cmc333333) [inline regparser/web/jobs/models.py]: Something to consider here -- you might not want to track all these fields in a separate model. It limits reusability; for example, we can't reuse this model to represent a `proposal_pipeline` job or a simpler `write_to` job. You can also pull this information from the `job` in Redis (until that job's deleted).

File: regparser/web/jobs/utils.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    :arg int timeout: (Optional) How long (in seconds) the job should run
+        before being considered "lost".
+    :arg int result_ttl: (Optional) How long (in seconds) the key for the job
+        result should be stored.
+
+    :rtype: uuid4
+    :returns: The UUID for the job.
+    """
+    return django_rq.enqueue(run_eregs_command, args, timeout=60*30,
+                             result_ttl=-1)
```
Comment (cmc333333) [inline regparser/web/jobs/utils.py]: Did you want to use the params for `timeout` and `result_ttl` on line 6 here?

File: regparser/web/jobs/utils.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    :rtype: None
+    :returns: None
+    """
+    queue = django_rq.get_queue()
+    redis_job = queue.fetch_job(str(job_id))
+    if redis_job:
+        redis_job.delete()
+
+
+def add_redis_data_to_job_data(job_data):
```
Comment (cmc333333) [inline regparser/web/jobs/utils.py]: This feels a little fishy -- it's trying to synchronize with an asynchronous thing. I bet `rq` has some sort of life-cycle hooks that we could use instead.

File: regparser/web/jobs/urls.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from django.conf.urls import url
+from regparser.web.jobs import views
+from rest_framework.urlpatterns import format_suffix_patterns
+
+urlpatterns = [
+    url(r'^job/$', views.JobViewList.as_view()),
+    url(r'^job/(?P<job_id>[-a-z0-9]+)/$',
```
Comment (cmc333333) [inline regparser/web/jobs/urls.py]: I feel a little antsy about revealing the job id to the end user. What do you think about just using the `pk` for now?

File: regparser/web/jobs/views.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            or about why the job couldn't be added (status 400).
+        """
+        serialized = self.get_serializer(data=request.data)
+        serialized.is_valid(raise_exception=True)
+        eregs_args = [
+            "pipeline",
+            str(serialized.validated_data["cfr_title"]),
+            str(serialized.validated_data["cfr_part"]),
+            "./testing"
+        ]
```
Comment (cmc333333) [inline regparser/web/jobs/views.py]: Note that this doesn't use all of the request data yet

File: regparser/web/jobs/views.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        job = queue_eregs_job(eregs_args, timeout=60*30, result_ttl=-1)
+
+        # Paranoia--validate the values we provide:
+        job_id = job.id
+        for validator in serialized.get_fields()["job_id"].validators:
+            validator(job_id)
+        statusurl = status_url(job_id)
+        for validator in serialized.get_fields()["url"].validators:
+            validator(statusurl)
```
Comment (cmc333333) [inline regparser/web/jobs/views.py]: This feels clunky -- can we just run `serialized.is_valid()` again?

File: regparser/web/jobs/views.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        """
+        Adding the Refresh header here so that the browser does the
+        user-friendly thing of redirecting the user to the page for the
+        newly-created object, even though use of the Refresh header is
+        frowned upon in some circles.
+
+        Not using redirect via 302 or 303 so that non-browser users get the
+        201 status code they expect upon a successful POST.
+
+        I'm open to debate on this decision.
```
Comment (cmc333333) [inline regparser/web/jobs/views.py]: Hrm, yes. I want to think this through a bit, but the basic mantra is to start simple. Edge cases should be noted, but I wouldn't spend cycles worrying about them yet.

File: regparser/web/jobs/serializers.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+class ParsingJobSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = ParsingJob
+        fields = (
+            "cfr_title",
+            "cfr_part",
+            "clear_cache",
+            "destination",  # Unsure about whether this should accept user
+                            # input or be set by the system.
```
Comment (cmc333333) [inline regparser/web/jobs/serializers.py]: Definitely should be set by the system in the final product; all sorts of security implications if not.

File: regparser/web/jobs/utils.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
+from regparser.tasks import run_eregs_command
+import django_rq
+import settings
+
+
+def queue_eregs_job(args, timeout=60*30, result_ttl=-1):
```
Comment (cmc333333) [inline regparser/web/jobs/utils.py]: Heads up that a `result_ttl` of `-1` (i.e. never delete) is okay for demonstration purposes, but we'll want to change it before too long. `rq` more or less has a list associated with a key like `rq:active_jobs` that'll keep just keep growing; this list'll be returned for _every_ `queue` interaction in the future.

**Comment by cmc333333** [comment]
I added a handful of comments, but I'm okay merging this as is. Nice work!

File: regparser/web/jobs/utils.py (hunk 2 above)
Comment (tadhg-ohiggins) [inline regparser/web/jobs/utils.py]: Yes, good catch.

File: regparser/web/jobs/urls.py (hunk 4 above)
Comment (tadhg-ohiggins) [inline regparser/web/jobs/urls.py]: Heh, I had it that way in an earlier version and then simplified it, but am fine with changing it back.

File: regparser/web/jobs/views.py (hunk 6 above)
Comment (tadhg-ohiggins) [inline regparser/web/jobs/views.py]: I wasn't able to get that to work;  the magic provided by `rest_framework` seems to assume you're not going to want to validate anything you're providing yourself.

**Comment by cmc333333** [comment]
:+1: Merging now

**Pull request merged**